File: Leetcode/Minimum Distance to Type a Word Using Two Fingers/Minimum Distance to Type a Word Using Two Fingers.java
class Solution {
    public int minimumDistance(String word) {
        //dp array for each character
        int dp[] = new int[26]; 
        //Declaring ans and save variable which will contain the maximum distance that can be saved by using only one finger
        int ans = 0, save = 0;
        //Finding length of given word
        int n = word.length();
        //Looping through each character of given word
        for (int i = 0; i < n - 1; ++i) {
            //Finding b and c
            int b = word.charAt(i) - 'A';
            int c = word.charAt(i + 1) - 'A';
            //Looping through each character
            for (int a = 0; a < 26; ++a){
                //dp will be equal to maximum of current value and looped character and difference of the distance
                dp[b] = Math.max(dp[b], dp[a] + distance(b, c) - distance(a, c)); 
            }    
            //Updating save and ans
            save = Math.max(save, dp[b]);
            ans += distance(b, c);
        }
        return ans - save;

    }

    //Method to find distance
    private int distance(int x, int y) {
        return Math.abs(x / 6 - y / 6) + Math.abs(x % 6 - y % 6);
    }
}

File: Leetcode/Minimum Distance to Type a Word Using Two Fingers/Minimum Distance to Type a Word Using Two Fingers .cpp
#include <bits/stdc++.h>
using namespace std;

//defining auto to specify the type of the variable
auto op = [](auto& x, auto& y, char c) { x = (c -'A')/6; y = (c-'A')%6; };
class Solution {
public:
    
   //declaration for dp array 
   int dp[301][301];    
    int minimumDistance(string word) 
    {
        //setting values
        memset(dp, 0, sizeof(dp));
        
        //Finding the length of given word
        word.insert(0, 1,'0');
        
        //declaring required variables
        int x1, y1, x2, y2, x3, y3;
        
        //loop for finding characters
        for (int i=word.length()-2; i>=0; i--)
        {
            op(x1,y1,word[i]); 
            op(x3,y3,word[i+1]);
            
            //loop for finding maximum of current value and difference of the distance
            for (int j=i-1; j>=0; j--)
            {
                op(x2,y2,word[j]);
                int d1 = word[i] != '0' ? abs(x1-x3) + abs(y1-y3) : 0;
                int d2 = word[j] != '0' ? abs(x2-x3) + abs(y2-y3) : 0;
                dp[i][j] = min(dp[i+1][j] + d1,
                               dp[i][i+1] + d2);
                dp[j][i] = dp[i][j];
            }
        }
        
        //returning minimum distance
        return dp[1][0];
    }
};